File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    parent = [0 for _ in range(n + 1)]
    
    def Find(node): 
        if parent[node] != 0:
            return Find(parent[node])
        return node
    
    def Union(a, b): 
        a = Find(a)
        b = Find(b)
        
        if a < b:
            parent[b] = a
        else:
            parent[a] = b
    
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            if i != j:
                if computers[i - 1][j - 1] == 1:
                    if Find(i) != Find(j):
                        Union(i, j)
    
    answer = parent.count(0) - 1
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution(n, computers) {
	let graph = Array.from({ length: n }, () => new Set([]));
	for (let i = 0; i < n; i++) {
		for (let j = 0; j < n; j++) {
			if (i === j) continue;

			if (computers[i][j] === 1) {
				graph[i].add(j);
				graph[j].add(i);
			}
		}
	}
	graph = graph.map((item) => [...item]);

	let answer = 0;
	let visited = Array(n).fill(false);

	function BFS(node) {
		visited[node] = true;
		let queue = [node];

		while (queue.length) {
			let curNode = queue.shift();

			for (let x of graph[curNode]) {
				if (!visited[x]) {
					BFS(x);
				}
			}
		}
	}

	for (let i = 0; i < n; i++) {
		if (!visited[i]) {
			answer++;
			BFS(i);
		}
	}

	return answer;
}